stl/src/print.cpp

```diff
@@ -0,0 +1,152 @@
+        const int _Fd = _fileno(_File_stream);
+
+        switch (_Fd) {
+        // _fileno() returns -2 if _File_stream refers to either stdout or stderr and
+        // there is no associated output stream. In that case, there is also no associated
+        // console HANDLE. (This might happen, for instance, if a Win32 GUI application is
+        // being compiled with /SUBSYSTEM:WINDOWS.)
+        case -2:
+            return __std_unicode_console_retrieval_result{._Error = __std_win_error::_Not_supported};
+
+        case -1:
+            [[unlikely]] return __std_unicode_console_retrieval_result{._Error = __std_win_error::_Invalid_parameter};
+
+        default:
+            break;
+        }
+
+        const HANDLE _Console_handle = reinterpret_cast<HANDLE>(_get_osfhandle(_Fd));
+
+        if (_Console_handle == INVALID_HANDLE_VALUE) [[unlikely]] {
+            return __std_unicode_console_retrieval_result{._Error = __std_win_error::_Invalid_parameter};
+        }
+
+        // We can check if _Console_handle actually refers to a console or not by checking the
+        // return value of GetConsoleMode().
+        DWORD _Console_mode;
+        const bool _Is_unicode_console = (GetConsoleMode(_Console_handle, &_Console_mode) != 0);
+
+        if (!_Is_unicode_console) {
+            return __std_unicode_console_retrieval_result{._Error = __std_win_error::_File_not_found};
+        }
+
+        return __std_unicode_console_retrieval_result{
+            ._Console_handle = static_cast<__std_unicode_console_handle>(
+                reinterpret_cast<_STD underlying_type_t<__std_unicode_console_handle>>(_Console_handle))};
+    }
+} // unnamed namespace
+
+_EXTERN_C
+
+[[nodiscard]] _Success_(return._Error == __std_win_error::_Success) __std_unicode_console_retrieval_result
+    __stdcall __std_get_unicode_console_handle_from_file_stream(_In_ const __std_file_stream_pointer _Stream) noexcept {
+    if (_Stream == __std_file_stream_pointer::_Invalid) [[unlikely]] {
+        return __std_unicode_console_retrieval_result{._Error = __std_win_error::_Invalid_parameter};
+    }
+
+    FILE* const _File_stream_ptr = reinterpret_cast<FILE*>(_Stream);
+    return _Get_console_handle_from_file_stream(_File_stream_ptr);
+}
+
+_END_EXTERN_C
+
+namespace {
+    [[nodiscard]] _STD expected<_STD wstring, __std_win_error> __stdcall _Transcode_utf8_string(
+        const char* const _Str, const size_t _Str_size) noexcept {
+        // MultiByteToWideChar() fails if strLength == 0.
+        if (_Str_size == 0) [[unlikely]] {
+            return {};
+        }
+
+        // The C++ specifications for vprint_unicode() suggest that we replace invalid
+        // code units with U+FFFD. This is done automatically by MultiByteToWideChar(),
+        // so long as we do not use the MB_ERR_INVALID_CHARS flag.
+        const int32_t _Num_chars_required =
+            MultiByteToWideChar(CP_UTF8, 0, _Str, static_cast<int>(_Str_size), nullptr, 0);

```

Comment:
If `_Str_size` does not fit into an `int`, this will silently print less than the full number of characters. (also, this allocates absolutely massive strings when `_Str_size` is large). We may want to only convert N characters at a time where N is something like 8096.

Reply:
I'm not sure if the large allocation issue is something we can/should reliably avoid. Ideally, we would want to prevent things like interleaved console output in multithreaded code. I can think of two ways to achieve these results:

  1. Print the entire string in one call to `WriteConsoleW()`. This requires that we allocate enough memory for a single contiguous character array.
  2. Use a `std::mutex` or some other similar mechanism to ensure that only one thread calls `WriteConsoleW()` at once. If we want to reduce memory consumption by, e.g., using a single buffer for transcoding and printing only parts of the user's string at a time, then we'll also need to acquire that lock during transcoding. So, we would be trading performance for memory.
  
I'd like some suggestions as to how we should handle this before I resolve this issue. On a different note, I completely agree that the "larger than `int`" issue is a legitimate error, so I'll be fixing that.